src/librustc_mir/transform/print_type_sizes.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This pass implements instrumentation to gather the layout of every type.
+
+use rustc::session::{VariantSize};
+use rustc::traits::{Reveal};
+use rustc::ty::{self, Ty, TyCtxt};
+use rustc::ty::fold::{TypeFoldable};
+use rustc::ty::layout::{Layout};
+use rustc::mir::{Mir};
+use rustc::mir::transform::{MirPass, MirPassHook, MirSource, Pass};
+use rustc::mir::visit::Visitor;
+
+use std::collections::HashSet;
+
+pub struct GatherTypeSizesMir {
+    _hidden: (),
+}
+
+impl GatherTypeSizesMir {
+    pub fn new() -> Self {
+        GatherTypeSizesMir { _hidden: () }
+    }
+}
+
+impl Pass for GatherTypeSizesMir {
+}
+
+impl<'tcx> MirPassHook<'tcx> for GatherTypeSizesMir {
+    fn on_mir_pass<'a>(&mut self,
+                       tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                       src: MirSource,
+                       mir: &Mir<'tcx>,
+                       _pass: &Pass,
+                       _is_after: bool) {
+        debug!("on_mir_pass: {}", tcx.node_path_str(src.item_id()));
+        self.go(tcx, mir);
+    }
+}
+
+impl<'tcx> MirPass<'tcx> for GatherTypeSizesMir {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    src: MirSource, mir: &mut Mir<'tcx>) {
+        debug!("run_pass: {}", tcx.node_path_str(src.item_id()));
+        self.go(tcx, mir);
+    }
+}
+
+impl GatherTypeSizesMir {
+    fn go<'a, 'tcx>(&mut self,
+                    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    mir: &Mir<'tcx>) {
+        if tcx.sess.err_count() > 0 {
+            // compiling a broken program can obviously result in a
+            // broken MIR, so do not bother trying to process it.
+            return;
+        }
+
+        let mut visitor = TypeVisitor {
```

Comment:
Shouldn't we have a single `seen` list for the entire program? (A: this is just a cache for the deduplication in session?).